src/test/java/EchoServerTest.java

```diff
@@ -0,0 +1,54 @@
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.*;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.ArrayList;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class EchoServerTest {
+    private int port = 7777;
+    private Client echoClient;
+    private ServerSocket serverSocket;
+    private Server echoServer;
+
+    private ArrayList<String> out = new ArrayList<>();
+    private ArrayList<String> in = new ArrayList<>();
+
+    @BeforeEach
+    public void setup() throws IOException {
+        echoClient = new EchoClientSpy();
+        serverSocket = new ServerSocket(port);
+        Socket socket = new Socket("localhost", port);
+        echoClient.start(socket);
+        in.add(echoClient.sendMessage("Roger"));
+        in.add(echoClient.sendMessage("and"));
+        in.add(echoClient.sendMessage("."));
+        in.add(echoClient.sendMessage("If this gets returned, it's all gone wrong"));
+    }
+
+    @Test
+    public void serverEchoesInputUntilFullStop() {
+        echoServer = new EchoServerSpy(serverSocket, in, out);
```

Comment:
**thought**: this is good 👍 how might you be able to change the design so that you'd more closely follow the SRP + wouldn't need to have a spy Server? (spoiler, you'll still need _a_ fake test double somewhere)